pkg/skaffold/build/ordered.go

```diff
@@ -35,49 +35,51 @@ type ArtifactBuilder func(ctx context.Context, out io.Writer, artifact *latest.A
 
 // For testing
 var (
-	buffSize      = bufferedLinesPerArtifact
-	runInSequence = InSequence
+	buffSize = bufferedLinesPerArtifact
 )
 
-// InParallel builds a list of artifacts in parallel but prints the logs in sequential order.
-func InParallel(ctx context.Context, out io.Writer, tags tag.ImageTags, artifacts []*latest.Artifact, buildArtifact ArtifactBuilder, concurrency int) ([]Artifact, error) {
-	if len(artifacts) == 0 {
-		return nil, nil
-	}
-
-	if len(artifacts) == 1 || concurrency == 1 {
-		return runInSequence(ctx, out, tags, artifacts, buildArtifact)
-	}
-
-	var wg sync.WaitGroup
-	defer wg.Wait()
-
+// InOrder builds a list of artifacts in dependency order.
+// `concurrency` specifies the max number of builds that can run at any one time. If concurrency is 0, then it's set to the length of the `artifacts` slice.
+// Each artifact build runs in its own goroutine. It limits the max concurrency using a buffered channel like a semaphore.
+// At the same time, it uses the `artifactDAG` to model the artifacts dependency graph and to make each artifact build wait for its required artifacts' builds.
+func InOrder(ctx context.Context, out io.Writer, tags tag.ImageTags, artifacts []*latest.Artifact, buildArtifact ArtifactBuilder, concurrency int) ([]Artifact, error) {
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
 	results := new(sync.Map)
 	outputs := make([]chan string, len(artifacts))
 
-	if concurrency == 0 {
-		concurrency = len(artifacts)
-	}
-	sem := make(chan bool, concurrency)
+	scheduler := newScheduler(artifacts, concurrency)
+	var wg sync.WaitGroup
+	defer wg.Wait()
 
-	// Run builds in //
 	wg.Add(len(artifacts))
 	for i := range artifacts {
 		outputs[i] = make(chan string, buffSize)
 		r, w := io.Pipe()
 
-		// Run build and write output/logs to piped writer and store build result in
-		// sync.Map
-		go func(i int) {
-			sem <- true
-			runBuild(ctx, w, tags, artifacts[i], results, buildArtifact)
-			<-sem
-
+		// Create a goroutine for each element in acmSlice. Each goroutine waits on its dependencies to finish building.
+		// Because our artifacts form a DAG, at least one of the goroutines should be able to start building.
+		go func(a *latest.Artifact) {
+			// Run build and write output/logs to piped writer and store build result in sync.Map
+			scheduler.run(ctx, a,
+				func() (string, error) {
+					event.BuildInProgress(a.ImageName)
+					return getBuildResult(ctx, w, tags, a, buildArtifact)
+				},
+				func(tag string) {
+					event.BuildComplete(a.ImageName)
+					ar := Artifact{ImageName: a.ImageName, Tag: tag}
+					results.Store(ar.ImageName, ar)
+					w.Close()
+				},
+				func(err error) {
+					event.BuildFailed(a.ImageName, err)
+					results.Store(a.ImageName, err)
+					w.Close()
+				})
```

Comment:
All of this could happen in the scheduler.run function. Is there a reason why you are passing this as functions? The same function is executed for every artifact.

Reply:
I removed this.